genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/JobDirectoryLimitsServiceImpl.java

```diff
@@ -0,0 +1,125 @@
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.agent.execution.services.impl;
+
+import com.netflix.genie.agent.execution.services.JobDirectoryLimitsService;
+import com.netflix.genie.agent.execution.services.KillService;
+import com.netflix.genie.agent.properties.AgentProperties;
+import com.netflix.genie.agent.properties.JobDirectoryLimitsServiceProperties;
+import com.netflix.genie.common.internal.dtos.DirectoryManifest;
+import com.netflix.genie.common.internal.services.JobDirectoryManifestCreatorService;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.scheduling.TaskScheduler;
+import org.springframework.util.unit.DataSize;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Comparator;
+import java.util.Optional;
+import java.util.concurrent.ScheduledFuture;
+
+/**
+ * Implementation of {@link JobDirectoryLimitsService} that periodically checks on the size and number of files
+ * using the manifest creator, rather than looking at the actual files.
+ * This implementation is not thread safe and does not check for double start / double stop.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Slf4j
+class JobDirectoryLimitsServiceImpl implements JobDirectoryLimitsService {
+    private final KillService killService;
+    private final JobDirectoryManifestCreatorService manifestCreatorService;
+    private final TaskScheduler taskScheduler;
+    private final JobDirectoryLimitsServiceProperties properties;
+    private ScheduledFuture<?> scheduledCheck;
+
+    JobDirectoryLimitsServiceImpl(
+        final KillService killService,
+        final JobDirectoryManifestCreatorService manifestCreatorService,
+        final TaskScheduler taskScheduler,
+        final AgentProperties agentProperties) {
+        this.killService = killService;
+        this.manifestCreatorService = manifestCreatorService;
+        this.taskScheduler = taskScheduler;
+        this.properties = agentProperties.getJobDirectoryLimitsService();
+    }
+
+    @Override
+    public void start(final Path jobDirectory) {
+        this.scheduledCheck = this.taskScheduler.scheduleAtFixedRate(
+            () -> this.checkFilesSize(jobDirectory),
+            this.properties.getCheckInterval()
+        );
+
+    }
+
+    @Override
+    public void stop() {
+        if (this.scheduledCheck != null) {
+            this.scheduledCheck.cancel(true);
+        }
+    }
+
+    private void checkFilesSize(final Path jobDirectory) {
+        final DirectoryManifest manifest;
+        try {
+            manifest = this.manifestCreatorService.getDirectoryManifest(jobDirectory);
```

Comment:
Rather than this can we not just use the same manifest we're periodically sending to the server?

Reply:
The creator service internally caches and recomputes periodically.
So this call does not force the creation of a new manifest from scratch.

Reply:
👍